### PR TITLE
add StreamState and rework HTTP2 stream state handling

### PR DESCRIPTION
**Pull request by geoffkizer**
Add the StreamState enumeration to explicitly track the stream state. Also, detect some protocol errors like sending DATA before HEADERS, or sending HEADERS twice.

Rework the way we use TaskCompletionSource to notify waiters that data has arrived on the stream. This should contribute to both #31313 and other planned perf work.

Add some tests and simplify some existing test code

@dotnet/ncl @stephentoub

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 1]
```diff
@@ -124,6 +114,11 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
             {
                 // TODO: ISSUE 31309: Optimize HPACK static table decoding
 
+                if (_state != StreamState.ExpectingHeaders)
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Can you document in a comment when SyncObject is needed / what state it protects?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 2]
```diff
@@ -163,12 +158,33 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
...
+                    {
+                        waiterTaskSource = _waiterTaskSource;
+                        _waiterTaskSource = null;
+                    }
+                }
+
+                if (waiterTaskSource != null)
+                {
+                    waiterTaskSource.SetResult(true);
+                }
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Nit: `waiterTaskSource?.SetResult(true);`

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 3]
```diff
@@ -191,64 +210,118 @@ public void OnResponseData(ReadOnlySpan<byte> buffer, bool endStream)
...
+                        waiterTaskSource = _waiterTaskSource;
+                        _waiterTaskSource = null;
                     }
                 }
 
-                if (readDataAvailable != null)
+                if (waiterTaskSource != null)
                 {
-                    readDataAvailable.SetResult(true);
+                    waiterTaskSource.SetResult(true);
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Same nit

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 4]
```diff
@@ -163,12 +158,33 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
...
+                    _state = endStream ? StreamState.Complete : StreamState.ExpectingData;
+
+                    if (_waiterTaskSource != null)
+                    {
+                        waiterTaskSource = _waiterTaskSource;
+                        _waiterTaskSource = null;
+                    }
+                }
+
+                if (waiterTaskSource != null)
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: This new pattern is much nicer / simpler / easier to prove is correct from a threading perspective. Thanks.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 5]
```diff
@@ -191,64 +210,118 @@ public void OnResponseData(ReadOnlySpan<byte> buffer, bool endStream)
...
+                        waiterTaskSource = _waiterTaskSource;
+                        _waiterTaskSource = null;
+                    }
+                }
 
-                    if (!_responseHeadersAvailable.Task.IsCompleted)
+                if (waiterTaskSource != null)
+                {
+                    waiterTaskSource.SetResult(true);
+                }
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Same nit

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 6]
```diff
@@ -191,64 +210,118 @@ public void OnResponseData(ReadOnlySpan<byte> buffer, bool endStream)
...
+                    {
+                        Debug.Assert(_waiterTaskSource == null);
+                        _waiterTaskSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+                        return (_waiterTaskSource.Task, false);
+                    }
+                    else if (_state == StreamState.ExpectingData)
+                    {
+                        return (null, false);
                     }
                     else
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Nit: use a `switch (_state)` ?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 7]
```diff
@@ -191,64 +210,118 @@ public void OnResponseData(ReadOnlySpan<byte> buffer, bool endStream)
...
                     }
                 }
             }
 
+            public async Task ReadResponseHeadersAsync()
+            {
+                // Wait for response headers to be read.
+                Task waiterTask;
+                bool emptyResponse;
+                (waiterTask, emptyResponse) = TryEnsureHeaders();
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Nit:
```C#
(Task waiterTask, bool emptyResponse) = TryEnsureHeaders();
```

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs [hunk 8]
```diff
@@ -191,64 +210,118 @@ public void OnResponseData(ReadOnlySpan<byte> buffer, bool endStream)
...
+                if (waiterTask != null)
+                {
+                    await waiterTask;
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: .ConfigureAwait(false)

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs (hunk 1 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Actually, I think I'm going to just always ensure we take the SyncObject lock, here and elsewhere.

The reason for not locking here in the first place was that we shouldn't have returned from SendAsync yet, and thus there should be no possibility for concurrent access between user and Http2Connection here. But I'm not sure that's always and forever true, e.g. cancellation or other exception. The cost of locking is small, so it seems best to just do it consistently here.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs (hunk 8 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: I intentionally leave these off just to test you.

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs]: Makes sense, keep me on my toes, like pop quizzes in grade school.

**Comment by stephentoub** [comment]
I'll fix up my change that switches away from TCS and get that up as a PR.

**Pull request merged**